Drop String::fromUTF8() overloads taking in a raw pointer and a size

cdumez opened the pull request:
#### df2cc0871eec5de263ca9f20816ef2afcc7b359a
<pre>
Drop String::fromUTF8() overloads taking in a raw pointer and a size
<a href="https://bugs.webkit.org/show_bug.cgi?id=271945">https://bugs.webkit.org/show_bug.cgi?id=271945</a>

Reviewed by Darin Adler.

Drop String::fromUTF8() overloads taking in a raw pointer and a size, in favor
of the ones taking in spans.

* Source/JavaScriptCore/debugger/DebuggerScope.cpp:
(JSC::DebuggerScope::name const):
* Source/JavaScriptCore/jsc.cpp:
(stringFromUTF):
(JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/parser/Lexer.cpp:
(JSC::Lexer&lt;CharacterType&gt;::parseCommentDirectiveValue):
* Source/JavaScriptCore/profiler/ProfilerBytecode.cpp:
(JSC::Profiler::Bytecode::toJSON const):
* Source/JavaScriptCore/profiler/ProfilerBytecodeSequence.cpp:
(JSC::Profiler::BytecodeSequence::addSequenceProperties const):
* Source/JavaScriptCore/profiler/ProfilerBytecodes.cpp:
(JSC::Profiler::Bytecodes::toJSON const):
* Source/JavaScriptCore/profiler/ProfilerCompilation.cpp:
(JSC::Profiler::Compilation::toJSON const):
* Source/JavaScriptCore/profiler/ProfilerCompiledBytecode.cpp:
(JSC::Profiler::CompiledBytecode::toJSON const):
* Source/JavaScriptCore/profiler/ProfilerEvent.cpp:
(JSC::Profiler::Event::toJSON const):
* Source/JavaScriptCore/runtime/SamplingProfiler.cpp:
(JSC::SamplingProfiler::stackTracesAsJSON):
* Source/JavaScriptCore/tools/FunctionAllowlist.cpp:
(JSC::FunctionAllowlist::contains const):
* Source/WTF/wtf/StringPrintStream.cpp:
(WTF::StringPrintStream::tryToString):
(WTF::StringPrintStream::toString):
(WTF::StringPrintStream::toStringWithLatin1Fallback):
* Source/WTF/wtf/URL.cpp:
(WTF::decodeEscapeSequencesFromParsedURL):
* Source/WTF/wtf/URLParser.cpp:
(WTF::URLParser::percentDecode):
(WTF::URLParser::parseHostAndPort):
(WTF::URLParser::formURLDecode):
* Source/WTF/wtf/URLParser.h:
* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::createTemporaryZipArchive):
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::make8Bit):
(WTF::fromUTF8Impl):
(WTF::String::fromUTF8):
(WTF::String::fromUTF8ReplacingInvalidSequences):
(WTF::String::fromUTF8WithLatin1Fallback):
* Source/WTF/wtf/text/WTFString.h:
(WTF::String::fromUTF8): Deleted.
* Source/WebCore/Modules/applepay/cocoa/PaymentCocoa.mm:
(WebCore::convert):
* Source/WebCore/Modules/async-clipboard/ClipboardItemBindingsDataSource.cpp:
(WebCore::ClipboardItemBindingsDataSource::ClipboardItemTypeLoader::dataAsString const):
* Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDMSessionClearKey.cpp:
(WebCore::CDMSessionClearKey::update):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCDataChannelHandler.cpp:
(WebCore::LibWebRTCDataChannelHandler::OnMessage):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCPeerConnectionBackend.cpp:
(WebCore::LibWebRTCPeerConnectionBackend::doAddIceCandidate):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCUtils.h:
(WebCore::fromStdString):
* Source/WebCore/Modules/webauthn/cbor/CBORReader.cpp:
(cbor::CBORReader::readString):
* Source/WebCore/Modules/websockets/WebSocketExtensionParser.cpp:
(WebCore::WebSocketExtensionParser::consumeQuotedString):
* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::trimToValidUTF8Length1024):
* Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.cpp:
(PAL::CryptoDigest::toHexString):
* Source/WebCore/editing/cocoa/WebContentReaderCocoa.mm:
(WebCore::extractMarkupAndArchive):
(WebCore::sanitizeMarkupWithArchive):
* Source/WebCore/html/parser/AtomHTMLToken.h:
(WebCore::AtomHTMLToken::AtomHTMLToken):
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::HTMLTreeBuilder::ExternalCharacterTokenBuffer::makeString const):
* Source/WebCore/inspector/agents/InspectorNetworkAgent.cpp:
(WebCore::buildObjectForResourceRequest):
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCDTMFSenderBackend.cpp:
(WebCore::toWTFString):
* Source/WebCore/platform/mock/RTCDataChannelHandlerMock.cpp:
(WebCore::RTCDataChannelHandlerMock::sendStringData):
* Source/WebCore/platform/network/HTTPParsers.cpp:
(WebCore::parseHTTPHeader):
* Source/WebCore/platform/sql/SQLiteFileSystem.cpp:
(WebCore::SQLiteFileSystem::computeHashForFileName):
* Source/WebCore/platform/sql/SQLiteStatement.cpp:
(WebCore::SQLiteStatement::columnValue):
(WebCore::SQLiteStatement::columnText):
* Source/WebCore/xml/parser/XMLDocumentParser.cpp:
(WebCore::XMLDocumentParser::updateLeafTextNode):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::toString):

Canonical link: <a href="https://commits.webkit.org/276889@main">https://commits.webkit.org/276889@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/59c1ff69b8010ab5aa3984233f0097793cd9c285

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46061 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25191 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48646 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48732 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42101 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48368 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29510 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22630 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37669 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46639 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22284 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39743 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18859 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19683 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40840 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4105 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/39294 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42382 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41190 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50523 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/45535 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21057 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17577 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/44843 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22357 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43727 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22716 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/52686 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6415 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22051 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10777 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/text/WTFString.cpp:
```diff
@@ -477,54 +477,46 @@ void String::convertTo16Bit()
-        return String();
-
-    if (!length)
+    if (string.empty())
```
This removes the ability to create a null string. Is that OK?

cdumez replied:
Historically, the fromUTF8() that took a span always returned the empty string when getting an empty span.
The fromUTF8() overloads that take null terminated strings (via a single pointer) do still return the null string when getting a null pointer (check is at the call site of this function).

I have tried to differentiate null vs empty in the span case but it introduced crashes in JSC code (which passes spans with null pointers but expects an empty string back).

As it stands, there are no regressions on tests so this is likely OK.

darinadler replied:
A little concerned for the future that `String::fromUTF8` differs from the `String` constructor in this respect.

cdumez replied:
Ok. I can look into updating the JSC call sites instead.

darinadler replied:
I don’t think it’s urgent to do this; maybe we want to go the opposite direction and stop creating null strings implicitly?

darinadler commented on Source/WebCore/bridge/IdentifierRep.cpp:
```diff
@@ -89,9 +89,9 @@ IdentifierRep* IdentifierRep::get(const char* name)
+        return nullptr;
   
-    String string = String::fromUTF8WithLatin1Fallback(name, strlen(name));
+    String string = String::fromUTF8WithLatin1Fallback({ name, strlen(name) });
```
Don't we have a helper for this length span thing?

cdumez replied:
Oh, indeed. Will reuse.

cdumez committed:
Drop String::fromUTF8() overloads taking in a raw pointer and a size
https://bugs.webkit.org/show_bug.cgi?id=271945

Reviewed by Darin Adler.

Drop String::fromUTF8() overloads taking in a raw pointer and a size, in favor
of the ones taking in spans.

* Source/JavaScriptCore/debugger/DebuggerScope.cpp:
(JSC::DebuggerScope::name const):
* Source/JavaScriptCore/jsc.cpp:
(stringFromUTF):
(JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/parser/Lexer.cpp:
(JSC::Lexer<CharacterType>::parseCommentDirectiveValue):
* Source/JavaScriptCore/profiler/ProfilerBytecode.cpp:
(JSC::Profiler::Bytecode::toJSON const):
* Source/JavaScriptCore/profiler/ProfilerBytecodeSequence.cpp:
(JSC::Profiler::BytecodeSequence::addSequenceProperties const):
* Source/JavaScriptCore/profiler/ProfilerBytecodes.cpp:
(JSC::Profiler::Bytecodes::toJSON const):
* Source/JavaScriptCore/profiler/ProfilerCompilation.cpp:
(JSC::Profiler::Compilation::toJSON const):
* Source/JavaScriptCore/profiler/ProfilerCompiledBytecode.cpp:
(JSC::Profiler::CompiledBytecode::toJSON const):
* Source/JavaScriptCore/profiler/ProfilerEvent.cpp:
(JSC::Profiler::Event::toJSON const):
* Source/JavaScriptCore/runtime/SamplingProfiler.cpp:
(JSC::SamplingProfiler::stackTracesAsJSON):
* Source/JavaScriptCore/tools/FunctionAllowlist.cpp:
(JSC::FunctionAllowlist::contains const):
* Source/WTF/wtf/StringPrintStream.cpp:
(WTF::StringPrintStream::tryToString):
(WTF::StringPrintStream::toString):
(WTF::StringPrintStream::toStringWithLatin1Fallback):
* Source/WTF/wtf/URL.cpp:
(WTF::decodeEscapeSequencesFromParsedURL):
* Source/WTF/wtf/URLParser.cpp:
(WTF::URLParser::percentDecode):
(WTF::URLParser::parseHostAndPort):
(WTF::URLParser::formURLDecode):
* Source/WTF/wtf/URLParser.h:
* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::createTemporaryZipArchive):
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::make8Bit):
(WTF::fromUTF8Impl):
(WTF::String::fromUTF8):
(WTF::String::fromUTF8ReplacingInvalidSequences):
(WTF::String::fromUTF8WithLatin1Fallback):
* Source/WTF/wtf/text/WTFString.h:
(WTF::String::fromUTF8): Deleted.
* Source/WebCore/Modules/applepay/cocoa/PaymentCocoa.mm:
(WebCore::convert):
* Source/WebCore/Modules/async-clipboard/ClipboardItemBindingsDataSource.cpp:
(WebCore::ClipboardItemBindingsDataSource::ClipboardItemTypeLoader::dataAsString const):
* Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDMSessionClearKey.cpp:
(WebCore::CDMSessionClearKey::update):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCDataChannelHandler.cpp:
(WebCore::LibWebRTCDataChannelHandler::OnMessage):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCPeerConnectionBackend.cpp:
(WebCore::LibWebRTCPeerConnectionBackend::doAddIceCandidate):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCUtils.h:
(WebCore::fromStdString):
* Source/WebCore/Modules/webauthn/cbor/CBORReader.cpp:
(cbor::CBORReader::readString):
* Source/WebCore/Modules/websockets/WebSocketExtensionParser.cpp:
(WebCore::WebSocketExtensionParser::consumeQuotedString):
* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::trimToValidUTF8Length1024):
* Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.cpp:
(PAL::CryptoDigest::toHexString):
* Source/WebCore/editing/cocoa/WebContentReaderCocoa.mm:
(WebCore::extractMarkupAndArchive):
(WebCore::sanitizeMarkupWithArchive):
* Source/WebCore/html/parser/AtomHTMLToken.h:
(WebCore::AtomHTMLToken::AtomHTMLToken):
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::HTMLTreeBuilder::ExternalCharacterTokenBuffer::makeString const):
* Source/WebCore/inspector/agents/InspectorNetworkAgent.cpp:
(WebCore::buildObjectForResourceRequest):
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCDTMFSenderBackend.cpp:
(WebCore::toWTFString):
* Source/WebCore/platform/mock/RTCDataChannelHandlerMock.cpp:
(WebCore::RTCDataChannelHandlerMock::sendStringData):
* Source/WebCore/platform/network/HTTPParsers.cpp:
(WebCore::parseHTTPHeader):
* Source/WebCore/platform/sql/SQLiteFileSystem.cpp:
(WebCore::SQLiteFileSystem::computeHashForFileName):
* Source/WebCore/platform/sql/SQLiteStatement.cpp:
(WebCore::SQLiteStatement::columnValue):
(WebCore::SQLiteStatement::columnText):
* Source/WebCore/xml/parser/XMLDocumentParser.cpp:
(WebCore::XMLDocumentParser::updateLeafTextNode):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::toString):

Canonical link: https://commits.webkit.org/276889@main
